### PR TITLE
Add CUDA graph rewriting

### PR DESCRIPTION
**Pull request by vinx13**

### CODE REVIEW

File: include/tvm/runtime/container/adt.h [hunk 1]
```diff
@@ -136,7 +136,7 @@ class ADT : public ObjectRef {
    */
   template <typename... Args>
   static ADT Tuple(Args&&... args) {
-    return ADT(0, std::forward<Args>(args)...);
+    return ADT(0, {std::forward<Args>(args)...});
```
Comment (tqchen) [inline include/tvm/runtime/container/adt.h]: would be great to first keep ADT inact. We should also consider moving to Array later

Comment (vinx13) [reply include/tvm/runtime/container/adt.h]: it isn't changing ADT. Without brackets in some cases it doesn't correctly call into the constructor that accepts initializer list https://github.com/mlc-ai/relax/blob/a8f562ed2bfdf12486863e4b6f553e0e4b9d8f12/include/tvm/runtime/container/adt.h#L110
I can also send a PR upstream separately for this

File: src/relax/backend/vm/codegen_vm.cc [hunk 2]
```diff
@@ -514,6 +522,39 @@ class CodeGenVM : public ExprFunctor<Instruction::Arg(const Expr&)> {
...
+  Instruction::Arg EmitCUDAGraphEndCapture(const CallNode* call) {
+    size_t dst_register = NewRegister();
+    builder_->EmitCall("vm.builtin.cuda_graph_end_capture", {}, dst_register);
+    return Instruction::Arg(Instruction::kRegister, dst_register);
+  }
+
+  Instruction::Arg EmitCUDAGraphLaunch(const CallNode* call) {
+    size_t dst_register = NewRegister();
+    builder_->EmitCall("vm.builtin.cuda_graph_launch", ConvertArgs(GetRef<Call>(call)),
+                       dst_register);
```
Comment (tqchen) [inline src/relax/backend/vm/codegen_vm.cc:555]: NOTE after we bring back let us move to https://github.com/tlc-pack/relax/blob/relax/src/relax/backend/vm/vm_builtin_lower.cc later

File: src/relax/backend/vm/codegen_vm.cc [hunk 3]
```diff
@@ -533,6 +574,10 @@ class CodeGenVM : public ExprFunctor<Instruction::Arg(const Expr&)> {
   const Op& assert_op_ = Op::Get("relax.assert_op");
   const Op& make_closure_op_ = Op::Get("relax.make_closure");
   const Op& invoke_closure_op_ = Op::Get("relax.invoke_closure");
+  const Op& get_captured_cuda_graph_op_ = Op::Get("relax.vm.builtin.get_captured_cuda_graph");
```
Comment (tqchen) [inline src/relax/backend/vm/codegen_vm.cc:577]: Can we try to lower directly to `call_builtin` here, so you don't need to have op lowering

Comment (tqchen) [reply src/relax/backend/vm/codegen_vm.cc:577]: see https://github.com/tlc-pack/relax/blob/relax/src/relax/backend/vm/vm_shape_lower.cc#L735

**Pull request closed without merging**